src/GovUk.Education.ExploreEducationStatistics.Publisher/Services/DataSetPublishingService.cs

```diff
@@ -0,0 +1,67 @@
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Publisher.Services.Interfaces;
+using Microsoft.EntityFrameworkCore;
+
+namespace GovUk.Education.ExploreEducationStatistics.Publisher.Services;
+
+public class DataSetPublishingService(
+    ContentDbContext contentDbContext,
+    PublicDataDbContext publicDataDbContext
+    ) : IDataSetPublishingService
+{
+    public async Task PublishDataSets(IEnumerable<Guid> releaseVersionIds)
+    {
+        // Find all Data Files associated with the given ReleaseVersions.
+        var dataFileIds = await contentDbContext
+            .ReleaseFiles
+            .Where(rf => releaseVersionIds.Contains(rf.ReleaseVersionId) && rf.File.Type == FileType.Data)
+            .Select(rf => rf.FileId)
+            .ToListAsync();
+
+        // Find all DataSets whose current Draft DataSetVersions reference any of the Release Versions' Data Files. 
+        var releaseVersionDataSets = await publicDataDbContext
+            .DataSets
+            .AsNoTracking()
+            .Include(dataSet => dataSet.LatestLiveVersion)
+            .Include(dataSet => dataSet.LatestDraftVersion)
+            .Where(dataSet => dataSet.LatestDraftVersion != null
+                              && dataFileIds.Contains(dataSet.LatestDraftVersion.CsvFileId))
+            .ToListAsync();
+
+        // For each DataSet due to have a Draft DataSetVersion go live with the given Release Versions, publish the
+        // Draft version, and ensure that the overarching DataSet is also published.
+        releaseVersionDataSets
+            .ForEach(dataSet => 
+            {
+                var currentDraftVersion = dataSet.LatestDraftVersion;
+                currentDraftVersion!.Status = DataSetVersionStatus.Published;
+                currentDraftVersion.Published = DateTimeOffset.UtcNow;
+
+                // Mark the overarching DataSet as Published, and set a Published date if one has not previously been
+                // set.
+                dataSet.Status = DataSetStatus.Published;
+                dataSet.Published ??= currentDraftVersion.Published;
+                
+                // Set the newly published DataSetVersion as the overarching DataSet's LatestLiveVersion,
+                // and unset it from being the latest Draft version.
+                dataSet.LatestLiveVersion = currentDraftVersion;
+                dataSet.LatestLiveVersionId = currentDraftVersion.Id;
+                dataSet.LatestDraftVersion = null;
+                dataSet.LatestDraftVersionId = null;
+                
+                publicDataDbContext.DataSets.Update(dataSet);
+                publicDataDbContext.DataSetVersions.Update(currentDraftVersion);
+            });
```

Comment:
There's an additional change required here, where we need to find datasets which have latest live versions that haven't just transitioned from draft to live via this process, because the dataset hasn't been altered for this release version being published, and update those versions ReleaseFile id's to point to the ReleaseFile id's which correspond with this release version.

This means that when retrieving a dataset's latest live version and looking at it's release file, you can always work out what the _latest_ release version is.

Reply:
I can't see any changes related to this. Do we need to discuss it further?

Reply:
Nope, I hadn't gotten around to it before.  It's in now if you wanna take a look.

Just FYI you'll see in the tests that there is the possibility that in a chain of Release amendments:

* ReleaseVersion 1 - "my-data-file.csv" is added and becomes the basis for DataSetVersion 1
* ReleaseVersion 2 - "my-data-file.csv" is retained in this amendment.  When this one is published, DataSetVersion 1 has its ReleaseFileId updated to reflect the ReleaseFile.Id for ReleaseVersion 2.
* ReleaseVersion 3 - "my-data-file.csv" is removed in this amendment.  When this one is published, the ReleaseFileId for DataSetVersion 1 is left unchanged, so it continues to point at ReleaseVersion 2 (which is no longer the latest live ReleaseVersion for this Release series.

To me this feels like we might err into bug territory if this is not expected behaviour, as you now have a DataSetVersion that is pointing at a no longer visible ReleaseVersion, but mabe you could give me your thoughts on that and what was being achieved by switching from CsvFileId to ReleaseFileId?

Cheers!

Reply:
I think though that if we had rules to say that a file that was the basis for a published DataSetVersion cannot be removed during an amendment, we'd be OK.

Reply:
In response to my question:

```
If a Data File has been uploaded to a Release and then used by a Public API Data Set, can it then subsequently be removed via a Release amendment?  I.e. the Data File is no longer visible on any published Release
```

the answer for now is:

```
For MVP, you won't be able to remove the dataset in that scenario
```

So I think that as it stands right now the test case where I remove a Data File in an amendment, when that Data File has already been chosen to be used for an API data set is maybe not valid for MVP, but could be beyond.  So it's a question of if I remove that test scenario for now or not.